src/material.cpp

```diff
@@ -5,55 +5,53 @@
 #include "json.h"
 #include "translations.h"
 #include "generic_factory.h"
+#include "item_factory.h"
 
 #include <string>
 #include <map>
 
+template<>
+const string_id<material_type> string_id<material_type>::NULL_ID( "null", 0 );
+
 namespace
 {
-std::map<material_id, material_type> _all_materials;
+
+generic_factory<material_type> material_data( "material", "ident" );
+
 } // namespace
 
 template<>
 bool string_id<material_type>::is_valid() const
 {
-    return _all_materials.count( *this ) > 0;
+    return material_data.is_valid( *this );
 }
 
 template<>
 const material_type &string_id<material_type>::obj() const
 {
-    const auto found = _all_materials.find( *this );
-    if( found == _all_materials.end() ) {
-        debugmsg( "Tried to get invalid material: %s", c_str() );
-        static const material_type null_material{};
-        return null_material;
-    }
-    return found->second;
-}
-
-material_type::material_type()
-{
-    _ident = material_id( "null" );
-    _name = "null";
-    _salvaged_into = itype_id( "null" );
-    _repaired_with = itype_id( "" );
-    _bash_resist = 0;
-    _cut_resist = 0;
-    _bash_dmg_verb = _( "damages" );
-    _cut_dmg_verb = _( "damages" );
-    _dmg_adj[0] = _( "lightly damaged" );
-    _dmg_adj[1] = _( "damaged" );
-    _dmg_adj[2] = _( "very damaged" );
-    _dmg_adj[3] = _( "thoroughly damaged" );
-    _acid_resist = 0;
-    _elec_resist = 0;
-    _fire_resist = 0;
-    _chip_resist = 0;
-    _density = 1;
-    _soft = false;
+    return material_data.obj( *this );
 }
 
+material_type::material_type() :
+    id( NULL_ID ),
+    _name(),
+    _salvaged_into( itype_id( "null" ) ),
+    _repaired_with( itype_id( "null" ) ),
+    _bash_resist( 0 ),
```

Comment:
> This later aids determination and usage of default values

That doesn't make much sense here since most of them are mandatory (at least for now) and hence cannot be defaulted, but on the other hand, there's only one constructor and several lines of code can be saved. I'll do that.